### DISCUSSION TITLE
ue4 客户端可以注册，但login总是timeout

### DISCUSSION THREAD

**Discussion by hujuly** [kbengine/kbengine#1569] [Q&A] [answered]
客户端我使用ue4开发，做了简单的登录和注册界面逻辑。
![1659979875661](https://user-images.githubusercontent.com/3659093/183478238-bd2ac4db-dd70-454a-b6a8-8ebb3dd57053.png)

服务器端，注册用户都可以成功，数据库里会有新用户记录。
![1659979931595](https://user-images.githubusercontent.com/3659093/183478407-e0b4d436-f4c0-42b6-b41e-4a88f0bc45b5.png)

但是login就会出现日志 pendingLoginMgr::add: username, size=1，接着就显示超时PendingLoginMgr::process:[username] is timeout, currsize=0.请问这是什么原因造成的？我应该如何调试？初学者，求详细点的帮助，谢谢！
![1659979958729](https://user-images.githubusercontent.com/3659093/183478476-3f6f5f33-81ba-4245-84ee-8f58964d5277.png)

**Comment by hujuly** [comment, marked as answer]
在什么情况下会调用PendingLoginMgr？这个还是没搞懂

**Reply by hujuly** [reply]
搞定了，配置问题

**Comment by riyueguanghui** [comment]
你好想问下，是哪个配置的问题

**Comment by ve-i-uj** [comment]
I guess that the Baseapp address or port in the configuration file was incorrect (root/baseapp/externalAddress | externalInterface | externalTcpPorts_min). It was able to connect to Loginapp, but Baseapp's address didn't point anywhere, so the client couldn't connect to Baseapp.

**Reply by riyueguanghui** [reply]
thank you for your answer，but i use another account is OK。

**Reply by ve-i-uj** [reply]
If it is not possible to log in under one account, but another account is OK, then you can try to delete all rows in the "kbe_entitylog" table (in the MySQL database). As far as I remember, these rows are not deleted when the engine is stopped forced. And then it cannot be possible to log in under the account that was active during force stop.

There can be many reasons, more information is needed: a return code from the server, logs.

**Reply by riyueguanghui** [reply]
loginapp logs: PendingLoginMgr::process:[username] is timeout. the server return error code is "login failed, err=SERVER_ERR_BUSY". I try to find account in "kbe_entitylog", but is't not exist. i think it's stuck at some step in the login process, and it don't run to the step of interfaces's function "onRequestAccountLogin".

**Reply by ve-i-uj** [reply]
It is necessary to check that Cellapp and Baseapp started normally: are Cellapp and Baseapp processes running? Are there errors from  Cellapp and Baseapp during startup in the logs? You also need to check if the onReadyForLogin callback returns 1.0  (in Baseapp kbemain.py)

**Reply by riyueguanghui** [reply]
This situation occurred in the online environment of the game, Cellapp and Baseapp is runing normally.  the onReadyForLogin would return 1.0 .

**Reply by ve-i-uj** [reply]
What is the value of the "flags" field in the "kbe_accountinfos" table of the problem account? (account must be verified by email?)

In the source code of KBEngine I found this error description:

操作过于繁忙(例如：在服务器前一次请求未执行完毕的情况下连续N次创建账号)。

Also the source code has only 3 locations for the "SERVER_ERR_BUSY" error. We can localize the source of the error by the log records that were before the error was sent.

https://github.com/kbengine/kbengine/blob/master/kbe/src/server/baseapp/baseapp.cpp#L3859

```
		INFO_MSG(fmt::format("Baseapp::loginBaseapp: user[{}] has entity({}).\n",
			accountName.c_str(), ptinfos->entityID));
```

https://github.com/kbengine/kbengine/blob/master/kbe/src/server/loginapp/loginapp.cpp#L356

```
		WARNING_MSG(fmt::format("Loginapp::_createAccount: pendingCreateMgr has {}, request create failed!\n", 
			accountName));
```

https://github.com/kbengine/kbengine/blob/master/kbe/src/server/loginapp/loginapp.cpp#L1090

```
	INFO_MSG(fmt::format("Loginapp::loginFailed: loginName={0} login failed. failedcode={1}, datas={2}.\n",
		loginName, SERVER_ERR_STR[failedcode], datas));
```

**Reply by riyueguanghui** [reply]
WARNING   loginapp01 47676 1011921682140  [2023-07-22 17:48:38 910] - Loginapp::onHello: client is not encrypted, addr=127.0.0.1:54844/0/0/0
   DEBUG   loginapp01 47676 1011921682140  [2023-07-22 17:48:38 910] - PendingLoginMgr::add: 1, size=1.
    INFO   loginapp01 47676 1011921682140  [2023-07-22 17:48:38 910] - Loginapp::login: new client[CLIENT_TYPE_BROWSER], loginName=1, datas={"openid":"1","openkey":"","channel":99999,"versions":"3.27","platform":"windows"}.
   DEBUG   loginapp01 47676 1011921682140  [2023-07-22 17:48:49 879] - Channel::handshake: websocket(127.0.0.1:54851/0/0/0) successfully!
    INFO   loginapp01 47676 1011921682140  [2023-07-22 17:48:49 880] - ServerApp::onHello: verInfo=1.3.7, scriptVerInfo=0.1.0, encryptedKey=None, addr:127.0.0.1:54851/0/0/0
 WARNING   loginapp01 47676 1011921682140  [2023-07-22 17:48:49 880] - Loginapp::onHello: client is not encrypted, addr=127.0.0.1:54851/0/0/0
    INFO   loginapp01 47676 1011921682140  [2023-07-22 17:48:49 880] - Loginapp::loginFailed: loginName=1 login failed. failedcode=SERVER_ERR_BUSY, datas=.
   DEBUG   loginapp01 47676 1011921682140  [2023-07-22 17:48:52 938] - Channel::handshake: websocket(127.0.0.1:54852/0/0/0) successfully!
    INFO   loginapp01 47676 1011921682140  [2023-07-22 17:48:52 939] - ServerApp::onHello: verInfo=1.3.7, scriptVerInfo=0.1.0, encryptedKey=None, addr:127.0.0.1:54852/0/0/0
 WARNING   loginapp01 47676 1011921682140  [2023-07-22 17:48:52 939] - Loginapp::onHello: client is not encrypted, addr=127.0.0.1:54852/0/0/0
    INFO   loginapp01 47676 1011921682140  [2023-07-22 17:48:52 939] - Loginapp::loginFailed: loginName=1 login failed. failedcode=SERVER_ERR_BUSY, datas=.
   DEBUG   loginapp01 47676 1011921682140  [2023-07-22 17:49:35 992] - Channel::handshake: websocket(127.0.0.1:54879/0/0/0) successfully!
    INFO   loginapp01 47676 1011921682140  [2023-07-22 17:49:35 993] - ServerApp::onHello: verInfo=1.3.7, scriptVerInfo=0.1.0, encryptedKey=None, addr:127.0.0.1:54879/0/0/0
 WARNING   loginapp01 47676 1011921682140  [2023-07-22 17:49:35 993] - Loginapp::onHello: client is not encrypted, addr=127.0.0.1:54879/0/0/0
    INFO   loginapp01 47676 1011921682140  [2023-07-22 17:49:35 994] - Loginapp::loginFailed: loginName=1 login failed. failedcode=SERVER_ERR_BUSY, datas=.
   DEBUG   loginapp01 47676 1011921682140  [2023-07-22 17:50:30 397] - Channel::handshake: websocket(127.0.0.1:54920/0/0/0) successfully!
    INFO   loginapp01 47676 1011921682140  [2023-07-22 17:50:30 398] - ServerApp::onHello: verInfo=1.3.7, scriptVerInfo=0.1.0, encryptedKey=None, addr:127.0.0.1:54920/0/0/0
 WARNING   loginapp01 47676 1011921682140  [2023-07-22 17:50:30 398] - Loginapp::onHello: client is not encrypted, addr=127.0.0.1:54920/0/0/0
   DEBUG   loginapp01 47676 1011921682140  [2023-07-22 17:50:30 398] - PendingLoginMgr::add: AAAA, size=2.
    INFO   loginapp01 47676 1011921682140  [2023-07-22 17:50:30 398] - Loginapp::login: new client[CLIENT_TYPE_BROWSER], loginName=AAAA, datas={"openid":"AAAA","openkey":"","channel":99999,"versions":"3.27","platform":"windows"}.
    INFO   loginapp01 47676 1011921682140  [2023-07-22 17:50:36 003] - ServerApp::onChannelTimeOut: Channel 127.0.0.1:54879/0/0/0 timeout!
   DEBUG   loginapp01 47676 1011921682140  [2023-07-22 17:50:36 967] - Channel::handshake: websocket(127.0.0.1:54925/0/0/0) successfully!
    INFO   loginapp01 47676 1011921682140  [2023-07-22 17:50:36 969] - ServerApp::onHello: verInfo=1.3.7, scriptVerInfo=0.1.0, encryptedKey=None, addr:127.0.0.1:54925/0/0/0
 WARNING   loginapp01 47676 1011921682140  [2023-07-22 17:50:36 969] - Loginapp::onHello: client is not encrypted, addr=127.0.0.1:54925/0/0/0
    INFO   loginapp01 47676 1011921682140  [2023-07-22 17:50:36 970] - Loginapp::loginFailed: loginName=AAAA login failed. failedcode=SERVER_ERR_BUSY, datas=.
   DEBUG   loginapp01 47676 1011921682140  [2023-07-22 17:50:38 924] - PendingLoginMgr::process: [1] is timeout, currsize=1.

some logs like this，i think it is output from https://github.com/kbengine/kbengine/blob/master/kbe/src/server/loginapp/loginapp.cpp#L1090

**Reply by ve-i-uj** [reply]
In [the first case](https://github.com/kbengine/kbengine/blob/master/kbe/src/server/baseapp/baseapp.cpp#L3859), the reason is the account is destroyed. It probably is the error in the python script of the Account.py.

In the [second case](https://github.com/kbengine/kbengine/blob/master/kbe/src/server/loginapp/loginapp.cpp#L356), the client may not have confirmed his account by email (this can be checked by the value of `kbe_accountinfos.flags` and the table `kbe_email_verification`).

If this is [the third case](https://github.com/kbengine/kbengine/blob/master/kbe/src/server/loginapp/loginapp.cpp#L1090), then I guess that the client sends several messages "Loginapp::login". And possibly receives several responses from the server (successful and error). Further, it depends on the client's logic (how it will react to an error message).

**Reply by ve-i-uj** [reply]
I think is the third case too. I guess that the error "SERVER_ERR_BUSY" is not a cause but a consequence. The user couldn't log in successfully and then click the button "log in" again and again. Thats why there is the error `SERVER_ERR_BUSY` in the logs. But the point is the why the user coundn't log in by the first time. It needs to find out by the logs what was before the error  `SERVER_ERR_BUSY`.

**Reply by ve-i-uj** [reply]
Is the client the Cocos2D demo?

**Reply by ve-i-uj** [reply]
If your client is based on the Cocos2D demo You must set the real ip addres [in this line](https://github.com/kbengine/kbengine_cocos2d_js_demo/blob/master/cocos2d-js-client/main.js#L57C13-L57C22) (not the localhost or `0.0.0.0`) otherwise you can correctly log in only from the PC where the server is running. From the other PC the game will be unreachable.

From the web browser you can request a login window as the request is handled by a web server (Nginx e.g.). But after sending the message "Loginapp::login" KBEngine sends the Baseapp address in the response. And the Baseapp address will be "localhost:20015" because this addres is set in the `args.ip`. So Baseapp will be unreachable you cannot connect to Baseapp and log in is failed.

So I think that is the reason of your problem.

**Reply by riyueguanghui** [reply]
I think is the second case ,  but i still haven't figured out what's causing it. It is not the client the Cocos2D demo, and the real ip addres is not localhost.

**Reply by ve-i-uj** [reply]
Can you reproduce the login issue? If you have the password of the problematic account, you can run [Wireshark] (https://www.wireshark.org/) on the client side and analyze network connections. Use port filters to capture Loginapp and Baseapp network traffic (`tcp.port == 20013 or tcp.port == 20015`). Try logging in with the problematic account, and then analyze the captured traffic. Captured traffic must have both connections. It is important that the connection to Baseapp exists (`tcp.port == 20015`).

![image](https://github.com/kbengine/kbengine/assets/6612371/00cc423f-8fa1-4451-ab32-0882b5f28a60)

**Comment by ve-i-uj** [comment]
> I think is the second case , but i still haven't figured out what's causing it. It is not the client the Cocos2D demo, and the real ip addres is not localhost.

What the value of `root/dbmgr/account_system/accountDefaultFlags` in the `kbengine.xml` file?

**Reply by riyueguanghui** [reply]
I find out the login issue.  
![image](https://github.com/kbengine/kbengine/assets/36831578/311dd02b-06bf-4c89-abd0-5e8bb5ed5a42)
if python script  function "onRequestAccountLogin“ throw exception and don't  execute the function "KBEngine.accountLoginResponse" ,
the loginName is always exist in reqAccountLogin_requests_.  so that it well be always login failure.
